src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,282 @@
+            /// reference count being set to zero.</para>
+            ///
+            /// <para>If the returned reference count box is <see langword="null"/>, then the target will also be
+            /// <see langword="null"/>.</para>
+            /// </returns>
+            private (T target, StrongBox<int> referenceCount) AtomicReadState()
+            {
+                // This value can be set in Dispose. A memory barrier is used to ensure _boxedReferenceCount is read
+                // after _instance (a stale read of _instance is not an error).
+                var target = _instance;
+                Interlocked.MemoryBarrier();
+                var referenceCount = Volatile.Read(ref _boxedReferenceCount);
+                if (referenceCount == null)
+                {
+                    return (null, null);
+                }
+                else
+                {
+                    return (target, referenceCount);
+                }
+            }
+
+            /// <summary>
+            /// Releases the current reference, causing the underlying object to be disposed if this was the last
+            /// reference.
+            /// </summary>
+            /// <remarks>
+            /// <para>After this instance is disposed, the <see cref="TryAddReference"/> method can no longer be used to
+            /// object a new reference to the target, even if other references to the target object are still in
+            /// use.</para>
+            /// </remarks>
+            public void Dispose()
+            {
+                var referenceCount = Interlocked.Exchange(ref _boxedReferenceCount, null);
+                if (referenceCount == null)
+                {
+                    // Already disposed; allow multiple without error.
+                    return;
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = Interlocked.Exchange(ref _instance, null);
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
+                if (decrementedValue == 0)
+                {
+                    instance.Dispose();
+                }
+#else
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    Debug.Assert(currentValue > 0, "Dispose should have protected itself against races.");
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue - 1, currentValue) == currentValue)
```

Comment:
If the compare exchange fails, you go back to the beginning of the `while(true)` and retry the read. This block is equivalent to the `Interlocked.Decrement` which is used in the alternate code path (now the only code path), except that it will assert rather than decrement the value if the loop is reached more than once for a given reference counted instance.